fastparquet/writer.py

```diff
@@ -930,197 +1078,155 @@ def write(filename, data, row_group_offsets=50000000,
     else:
-        raise ValueError('File scheme should be simple|hive, not', file_scheme)
+        # Case 'append=False'.
+        # Define 'index_cols' to be recorded in metadata.
+        if (write_index or write_index is None
+            and not isinstance(data.index, pd.RangeIndex)):
+            # Keep name(s) of index to metadata.
+            cols = set(data)
+            data = reset_row_idx(data)
+            index_cols = [c for c in data if c not in cols]
+        elif write_index is None and isinstance(data.index, pd.RangeIndex):
+            # write_index=None, range to metadata
+            index_cols = data.index
+        else:
+            # write_index=False
+            index_cols = []
+        # Initialize common metadata.
+        if str(has_nulls) == 'infer':
+            has_nulls = None
+        check_column_names(data.columns, partition_on, fixed_text,
+                           object_encoding, has_nulls)
+        ignore = partition_on if file_scheme != 'simple' else []
+        fmd = make_metadata(data, has_nulls=has_nulls, ignore_columns=ignore,
+                            fixed_text=fixed_text,
+                            object_encoding=object_encoding,
+                            times=times, index_cols=index_cols,
+                            partition_cols=partition_on)
+        if custom_metadata is not None:
+            kvm = fmd.key_value_metadata or []
+            kvm.extend(
+                [
+                    parquet_thrift.KeyValue(key=key, value=value)
+                    for key, value in custom_metadata.items()
+                ]
+            )
+            fmd.key_value_metadata = kvm
+
+        if file_scheme == 'simple':
+            # Case 'simple'
+            write_simple(filename, data, fmd,
+                         row_group_offsets=row_group_offsets,
+                         compression=compression, open_with=open_with,
+                         has_nulls=None, append=False, stats=stats)
+        else:
+            # Case 'hive', 'drill'
+            write_multi(filename, data, fmd,
+                        row_group_offsets=row_group_offsets,
+                        compression=compression, file_scheme=file_scheme,
+                        write_fmd=True, open_with=open_with,
+                        mkdirs=mkdirs, partition_on=partition_on,
+                        append=False, stats=stats)
 
 
 def find_max_part(row_groups):
     """
     Find the highest integer matching "**part.*.parquet" in referenced paths.
     """
-    paths = [c.file_path or "" for rg in row_groups for c in rg.columns]
-    # TODO: the following line should not be necessary
-    paths = [p.decode() if isinstance(p, bytes) else p for p in paths]
-    s = re.compile(r'.*part.(?P<i>[\d]+).parquet$')
-    matches = [s.match(path) for path in paths]
-    nums = [int(match.groupdict()['i']) for match in matches if match]
-    if nums:
-        return max(nums) + 1
+    pids = part_ids(row_groups)
```

Comment:
Ah, the source of the extra utility function. Is it really useful to pull it out of here, does it get called anywhere else?

Reply:
It is called in `ParquetFile._sort_part_names()`, where it uses the complementary info that are kept in the returned `dict`.

But I do agree, it seems to me we have several functions to handle partition info:
- directory path (partition values)
- file path (partition file id)
To query several type of info:
- list of partition file ids
- list of partition values
- max partition file id
- number of row groups per partition files
- mapping between partition file ids and row group id of the 1st row group the file contains

At present, it seems to me most of these are in `api.py`.